cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -621,11 +639,38 @@ Result<std::shared_ptr<FileWriter>> ParquetFileFormat::MakeWriter(
 
   auto parquet_options = checked_pointer_cast<ParquetFileWriteOptions>(options);
 
-  std::unique_ptr<parquet::arrow::FileWriter> parquet_writer;
-  ARROW_ASSIGN_OR_RAISE(parquet_writer, parquet::arrow::FileWriter::Open(
-                                            *schema, default_memory_pool(), destination,
-                                            parquet_options->writer_properties,
-                                            parquet_options->arrow_writer_properties));
+  std::unique_ptr<parquet::arrow::FileWriter> parquet_writer = NULLPTR;
+
+#ifdef PARQUET_REQUIRE_ENCRYPTION
+  std::shared_ptr<DatasetEncryptionConfiguration> dataset_encrypt_config =
+      GetDatasetEncryptionConfig();
+
+  if (dataset_encrypt_config != nullptr) {
+    auto file_encryption_prop =
+        dataset_encrypt_config->crypto_factory->GetFileEncryptionProperties(
+            *dataset_encrypt_config->kms_connection_config,
+            *dataset_encrypt_config->encryption_config, destination_locator.path,
+            destination_locator.filesystem);
+
+    auto writer_properties =
+        parquet::WriterProperties::Builder(*parquet_options->writer_properties)
+            .encryption(std::move(file_encryption_prop))
+            ->build();
+
+    ARROW_ASSIGN_OR_RAISE(
+        parquet_writer, parquet::arrow::FileWriter::Open(
+                            *schema, writer_properties->memory_pool(), destination,
+                            writer_properties, parquet_options->arrow_writer_properties));
+  }
+#endif
+
+  if (parquet_writer == NULLPTR) {
```

Comment:
```suggestion
  if (parquet_writer == nullptr) {
```

nit: it would be good to be consistent to use `nullptr` in the source file.